### PR TITLE
Feature/condo/sberdoma 207/add property import

### PR DESCRIPTION
**Pull request by Dimitreee**

### CODE REVIEW

File: apps/condo/domains/property/components/PropertyImport.tsx [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+    return [importData, progress, error, isImported, breakImport]
+}
+
+const ModalContext = React.createContext({ progress: 0, error: null, isImported: false })
+
+const getPropertyUploadInfoModalConfig = (onButtonClick) => {
+    return {
+        // TODO(Dimitreee):add translations
+        title: 'Импорт',
```
Comment (pahaz) [inline apps/condo/domains/property/components/PropertyImport.tsx:76]: Translation

File: apps/condo/domains/property/components/PropertyImport.tsx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+                            <>
+                                <Progress
+                                    format={(percent) => Math.floor(percent) + '%'}
+                                    percent={progress}
+                                    status={'active'}
+                                />
+                                <Alert
+                                    style={{ marginTop: 16 }}
+                                    // TODO(Dimitreee): add translations
+                                    message='Данные обрабатываются, не закрывайте вкладку браузера и не выключайте компьютер.'
```
Comment (pahaz) [inline apps/condo/domains/property/components/PropertyImport.tsx:92]: Translation

File: apps/condo/domains/property/components/PropertyImport.tsx [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+                                    type='info'
+                                />
+                            </>
+                        )
+                    }
+                }
+            </ModalContext.Consumer>
+        ),
+        okText: 'Прервать',
```
Comment (pahaz) [inline apps/condo/domains/property/components/PropertyImport.tsx:101]: Translation

File: apps/condo/domains/property/components/PropertyImport.tsx [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+            type: 'primary',
+            danger: true,
+        },
+    }
+}
+
+const getPropertyUploadSuccessModalConfig = () => {
+    return {
+        // TODO(Dimitreee):add translations
+        title: 'Импорт',
```
Comment (pahaz) [inline apps/condo/domains/property/components/PropertyImport.tsx:113]: Translation

File: apps/condo/domains/property/components/PropertyImport.tsx [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        closable: true,
+        content: (
+            <Alert
+                style={{ marginTop: 16 }}
+                // TODO(Dimitreee): add translations
+                message='Дома добавлены.'
+                type='success'
+            />
+        ),
+        okText: 'Продолжить',
```
Comment (pahaz) [inline apps/condo/domains/property/components/PropertyImport.tsx:123]: Translation

File: apps/condo/domains/property/components/PropertyImport.tsx [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+const getPropertyUploadSuccessModalConfig = () => {
+    return {
+        // TODO(Dimitreee):add translations
+        title: 'Импорт',
+        closable: true,
+        content: (
+            <Alert
+                style={{ marginTop: 16 }}
+                // TODO(Dimitreee): add translations
+                message='Дома добавлены.'
```
Comment (pahaz) [inline apps/condo/domains/property/components/PropertyImport.tsx:119]: Translation

File: apps/condo/domains/property/components/PropertyImport.tsx [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+            const config = getPropertyUploadErrorModalConfig()
+            activeModal.current = modal.error(config)
+        },
+    )
+
+    const handleUpload = useCallback((file) => {
+        destroyActiveModal()
+        const config = getPropertyUploadInfoModalConfig(() => {
+            breakImport()
+            props.onFinish()}
```
Comment (pahaz) [inline apps/condo/domains/property/components/PropertyImport.tsx:190]: Style

File: apps/condo/domains/property/components/PropertyImport.tsx [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        )
+        activeModal.current = modal.info(config)
+
+        importData(file.data)
+    }, [])
+
+    const organizationCreatorId = get(userOrganization, ['organization', 'createdBy', 'id'])
+    const userID = get(auth, ['user', 'id'])
+    const isOwner = organizationCreatorId === userID
```
Comment (pahaz) [inline apps/condo/domains/property/components/PropertyImport.tsx:199]: It's not true. You should check the Role permissions here. Employee.role.canManageProperties

File: apps/condo/domains/property/utils/PropertyImporter.ts [hunk 9]
```diff
@@ -0,0 +1,231 @@
+/*
+* На вход принимает excel файл
```
Comment (pahaz) [inline apps/condo/domains/property/utils/PropertyImporter.ts:2]: Cyrillic

File: apps/condo/domains/common/components/containers/FormTableExcelImport.tsx [hunk 10]
```diff
@@ -163,6 +167,9 @@ function ExcelExporterButton ({ columns, setExportedData }) {
         setExportedData(fromExData(tableState.data, tableState.mapping))
     }
 
+    const formatted_files = TABLE_UPLOAD_ACCEPT_FILES.map(function (x) { return '.' + x }).join(',')
```
Comment (pahaz) [inline apps/condo/domains/common/components/containers/FormTableExcelImport.tsx:170]: formattedFiles %)

File: apps/condo/domains/property/utils/PropertyImporter.ts [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+                minFloor: 1,
+                maxFloor: floors,
+            })
+
+            propertyUnitsMap.sections.push(sectionData)
+        }
+
+        const updatedSections = updateUnitsLabels(propertyUnitsMap)
+
+        // TODO: непонятно что делать с оставшимися после округления квартирами
```
Comment (pahaz) [inline apps/condo/domains/property/utils/PropertyImporter.ts:213]: cyrilic

File: apps/condo/domains/property/utils/PropertyImporter.ts [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+        }
+
+        const propertyUnitsMap = {
+            dv: 1,
+            type: BDataTypes.Building,
+            sections: [],
+        }
+
+        for (let currentSection = 0; currentSection < sections; currentSection++) {
+            const sectionName = `Подъезд №${currentSection + 1}`
```
Comment (pahaz) [inline apps/condo/domains/property/utils/PropertyImporter.ts:200]: translations

File: apps/condo/domains/property/utils/PropertyImporter.ts [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+
+    private isColumnsValid (row: TableRow): boolean {
+        const validColumns = ['address', 'units', 'sections', 'floors']
+        const normalizedTableColumns = row.map(({ value }) => {
+            if (typeof value === 'string') {
+                return value.toLowerCase()
+            }
+        })
+
+        // Проверяет корректность колонок файла
```
Comment (pahaz) [inline apps/condo/domains/property/utils/PropertyImporter.ts:151]: cyrilic

File: apps/condo/domains/property/utils/PropertyImporter.ts [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+        private addressApi: IAddressApi = new AddressApi(),
+    ) {}
+
+    // TODO(Dimitreee): remove any
+    public import (data: Array<TableRow>): Promise<any> {
+        this.tableData = data
+        const [columns, ...body] = this.tableData
+
+        if (!this.isColumnsValid(columns)) {
+            this.errorHandler(new Error('Таблица имеет некорректную структуру. Обратитесь к администратору за дополнительной информацией.'))
```
Comment (pahaz) [inline apps/condo/domains/property/utils/PropertyImporter.ts:51]: translation

**Comment by Dimitreee** [comment]
Rebase branch on master, remove direct DADATA api mentioning

File: apps/condo/domains/property/components/PropertyImport.tsx (hunk 8 above)
Comment (Dimitreee) [reply apps/condo/domains/property/components/PropertyImport.tsx:199]: going to use isAdmin

File: apps/condo/domains/common/constants/fileExtensions.js [hunk 15]
```diff
@@ -0,0 +1,5 @@
+export const TABLE_UPLOAD_ACCEPT_FILES = [
```
Comment (pahaz) [inline apps/condo/domains/common/constants/fileExtensions.js:1]: I think really this const will be used only in one place. And maybe it's better to place it near the usage (Optional)

File: apps/condo/domains/property/components/PropertyImport.tsx [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+
+    return [importData, progress, error, isImported, breakImport]
+}
+
+const ModalContext = React.createContext({ progress: 0, error: null, isImported: false })
+
+const getPropertyUploadInfoModalConfig = (intl, onButtonClick) => {
+    return {
+        // TODO(Dimitreee):add translations
+        title: intl.formatMessage({ id: 'property.Import' }),
```
Comment (pahaz) [inline apps/condo/domains/property/components/PropertyImport.tsx:77]: TitleMessage

File: apps/condo/domains/property/components/PropertyImport.tsx [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+                                    message={intl.formatMessage({ id: 'property.Processing' })}
+                                    type='info'
+                                />
+                            </>
+                        )
+                    }
+                }
+            </ModalContext.Consumer>
+        ),
+        okText: intl.formatMessage({ id: 'property.Break' }),
```
Comment (pahaz) [inline apps/condo/domains/property/components/PropertyImport.tsx:102]: OkMessage

File: apps/condo/domains/property/utils/PropertyImporter.ts [hunk 18]
```diff
@@ -0,0 +1,229 @@
+/*
+* Accepts an excel file as input
+* Parses it column by column
```
Comment (pahaz) [inline apps/condo/domains/property/utils/PropertyImporter.ts:3]:
```suggestion
* Parses the file columns
```

File: apps/condo/domains/property/utils/PropertyImporter.ts [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+* Validates address
+* Normalizes address
+* Checks the uniqueness of the address
+* Generates a checkerboard according to the specified parameters (If there are no parameters, it substitutes the default markup)
```
Comment (pahaz) [inline apps/condo/domains/property/utils/PropertyImporter.ts:7]:
```suggestion
* Generates a Property.map according to the specified parameters (If there are no parameters, it substitutes the default markup)
```

**Review by dkoviazin — APPROVED** [review]
Ok

**Pull request merged**